polkadot/xcm/src/v3/mod.rs

```diff
@@ -1440,15 +1453,39 @@ mod tests {
 	}
 
 	#[test]
-	fn decoding_fails_when_too_many_instructions() {
-		let small_xcm = Xcm::<()>(vec![ClearOrigin; 20]);
-		let bytes = small_xcm.encode();
+	fn decoding_limits() {
+		let max_xcm = Xcm::<()>(vec![ClearOrigin; MAX_INSTRUCTIONS_TO_DECODE as usize]);
+		let bytes = max_xcm.encode();
 		let decoded_xcm = Xcm::<()>::decode(&mut &bytes[..]);
 		assert!(matches!(decoded_xcm, Ok(_)));
 
-		let big_xcm = Xcm::<()>(vec![ClearOrigin; 64_000]);
+		let big_xcm = Xcm::<()>(vec![ClearOrigin; MAX_INSTRUCTIONS_TO_DECODE as usize + 1]);
 		let bytes = big_xcm.encode();
 		let decoded_xcm = Xcm::<()>::decode(&mut &bytes[..]);
 		assert!(matches!(decoded_xcm, Err(CodecError { .. })));
+
+		let nested_xcm = Xcm::<()>(vec![
+			DepositReserveAsset {
+				assets: All.into(),
+				dest: Here.into(),
+				xcm: max_xcm,
+			};
+			(MAX_INSTRUCTIONS_TO_DECODE / 2) as usize
+		]);
+		let bytes = nested_xcm.encode();
+		let decoded_xcm = Xcm::<()>::decode(&mut &bytes[..]);
+		assert!(matches!(decoded_xcm, Err(CodecError { .. })));
+
+		let even_more_nested_xcm = Xcm::<()>(vec![
+			DepositReserveAsset {
+				assets: All.into(),
+				dest: Here.into(),
+				xcm: nested_xcm,
+			};
+			(MAX_INSTRUCTIONS_TO_DECODE / 2) as usize
+		]);
+		let bytes = even_more_nested_xcm.encode();
+		let decoded_xcm = Xcm::<()>::decode(&mut &bytes[..]);
+		assert!(matches!(decoded_xcm, Err(CodecError { .. })));
 	}
```

Comment:
I had one local test in reserve, but looks likt it passes now :stuck_out_tongue:  
```rust
#[test]
fn decoding_respects_limit() {
	let layer0 = Xcm::<()>(vec![ClearOrigin; 64]);
	let layer1 = Xcm::<()>(vec![SetAppendix(layer0); 64]);
	let layer2 = Xcm::<()>(vec![SetAppendix(layer1); 16]);
	// We now have ~65K Instructions in here with an in-memory size of ~88 MB:
	let encoded = layer2.encode();
	assert_eq!(encoded.len(), 68657);

	// This should obviously not decode since the limit is 100:
	assert_eq!(MAX_INSTRUCTIONS_TO_DECODE, 100, "precondition");
	assert!(Xcm::<()>::decode(&mut &encoded[..]).is_err());
}